### PR TITLE
Migration/tabnote

### PR DESCRIPTION
**Pull request by rvilarl**
TabNote

### CODE REVIEW

File: src/tabnote.ts [hunk 1]
```diff
@@ -228,21 +240,21 @@ export class TabNote extends StemmableNote {
...
+  setStave(stave: Stave): this {
     super.setStave(stave);
-    this.context = stave.context;
+    this.setContext(stave.getContext());
 
     // Calculate the fret number width based on font used
     let i;
-    if (this.context) {
-      const ctx = this.context;
+    const ctx = this.getContext();
```
Comment (0xfe) [inline src/tabnote.ts:249]: You can remove a `getContext` call with:

```typescript
setStave(..) {
  super.setStave(stave);
  const ctx = stave.getContext();
  this.setContext(ctx);

  let i;
  ...
}
```

File: src/tabnote.ts [hunk 2]
```diff
@@ -228,21 +240,21 @@ export class TabNote extends StemmableNote {
...
 
   // Set the `stave` to the note
-  setStave(stave) {
+  setStave(stave: Stave): this {
     super.setStave(stave);
-    this.context = stave.context;
+    this.setContext(stave.getContext());
 
     // Calculate the fret number width based on font used
     let i;
```
Comment (0xfe) [inline src/tabnote.ts:248]: I think you can remove this, and move the `let` to the beginning of the loop below.

File: src/tabnote.ts [hunk 3]
```diff
@@ -266,13 +278,13 @@ export class TabNote extends StemmableNote {
...
-  getPositions() {
+  getPositions(): { str: string; fret: string }[] {
     return this.positions;
   }
 
   // Add self to the provided modifier context `mc`
-  addToModifierContext(mc) {
-    this.setModifierContext(mc);
+  addToModifierContext(mc: ModifierContext): this {
+    this.modifierContext = mc;
```
Comment (0xfe) [inline src/tabnote.ts:287]: Why not `this.setModifierContext`?

Comment (rvilarl) [reply src/tabnote.ts:287]: To let the compiler realize that `this.modifierContext` is not `undefined` avoiding the usage of `!`. 
It is a `protected` member and we save a function call which simply does the assignment.

File: src/tabnote.ts (hunk 1 above)
Comment (rvilarl) [reply src/tabnote.ts:249]: done

File: src/tabnote.ts (hunk 2 above)
Comment (rvilarl) [reply src/tabnote.ts]: done

**Comment by rvilarl** [comment]
Ready to merge!

**Comment by rvilarl** [comment]
@0xfe anything pending here?

File: src/tabnote.ts (hunk 3 above)
Comment (0xfe) [reply src/tabnote.ts:286]: Okay, that's a good reason. Looking at the codebase, it seems like `setModifierContext` doesn't do anything special, and it's only called in three places.

Can you remove `setModifierContext` entirely from `note.ts` and `modifier.ts`, and the references in `stavenote.ts`, `tabnote.ts`, and `modifiercontext.ts` too in this PR please? It's a good cleanup.

**Review by 0xfe — COMMENTED** [review]
One minor comment, rest of the change looks good.

File: src/tabnote.ts (hunk 3 above)
Comment (rvilarl) [reply src/tabnote.ts:286]: I have modified `StaveNote` to use the protected member. We cannot remove `setModifierContext` it is used by `ModifierContext` in `addMember`

**Comment by rvilarl** [comment]
I have modified `StaveNote` to use the protected member. We cannot remove `setModifierContext` it is used by `ModifierContext` in `addMember`
@0xfe ready to merge

File: src/tabnote.ts (hunk 3 above)
Comment (0xfe) [reply src/tabnote.ts:286]: Okay, yes, thanks.

File: src/tabnote.ts [hunk 4]
```diff
@@ -431,33 +449,40 @@ export class TabNote extends StemmableNote {
...
-  draw() {
-    this.checkContext();
+  draw(): void {
+    const ctx = this.checkContext();
 
     if (!this.stave) {
       throw new Vex.RERR('NoStave', "Can't draw without a stave.");
     }
 
+    if (!this.stem) {
```
Comment (0xfe) [inline src/tabnote.ts:474]: This seems wrong -- you can have tab notes without stems. Does this affect the tests?

File: src/tabnote.ts [hunk 5]
```diff
@@ -466,13 +491,13 @@ export class TabNote extends StemmableNote {
     this.stem.setNoteHeadXBounds(stem_x, stem_x);
 
     if (render_stem) {
```
Comment (0xfe) [inline src/tabnote.ts:493]: This is a better place to put a check for `this.stem`.

File: src/tabnote.ts (hunk 4 above)
Comment (rvilarl) [reply src/tabnote.ts]: No impact on tests. However I agree with you. I have changed the implementation and it also pass all the tests.

File: src/tabnote.ts (hunk 5 above)
Comment (rvilarl) [reply src/tabnote.ts]: Done this way, I am happy to see we agree! :)

**Comment by rvilarl** [comment]
@0xfe ready to merge!

**Comment by 0xfe** [comment]
lgtm

**Pull request merged**